README
NAME
    File::Listing - Parse directory listing

SYNOPSIS
     use File::Listing qw(parse_dir);
     $ENV{LANG} = "C";  # dates in non-English locales not supported
     foreach my $file (parse_dir(`ls -l`)) {
         my ($name, $type, $size, $mtime, $mode) = @$file;
         next if $type ne 'f'; # plain file
         #...
     }
 
     # directory listing can also be read from a file
     open my $listing, "zcat ls-lR.gz|";
     $dir = parse_dir($listing, '+0000');

DESCRIPTION
    This module exports a single function called "parse_dir", which can be
    used to parse directory listings.

FUNCTIONS
  parse_dir
     my $dir = parse_dir( $listing );
     my $dir = parse_dir( $listing, $time_zone );
     my $dir = parse_dir( $listing, $time_zone, $type );
     my $dir = parse_dir( $listing, $time_zone, $type, $error );
     my @files = parse_dir( $listing );
     my @files = parse_dir( $listing, $time_zone );
     my @files = parse_dir( $listing, $time_zone, $type );
     my @files = parse_dir( $listing, $time_zone, $type, $error );

    The first parameter ($listing) is the directory listing to parse. It can
    be a scalar, a reference to an array of directory lines or a glob
    representing a filehandle to read the directory listing from.

    The second parameter ($time_zone) is the time zone to use when parsing
    time stamps in the listing. If this value is undefined, then the local
    time zone is assumed.

    The third parameter ($type) is the type of listing to assume. Currently
    supported formats are 'unix', 'apache' and 'dosftp'. The default value
    is 'unix'. Ideally, the listing type should be determined automatically.

    The fourth parameter ($error) specifies how unparseable lines should be
    treated. Values can be 'ignore', 'warn' or a code reference. Warn means
    that the perl warn() function will be called. If a code reference is
    passed, then this routine will be called and the return value from it
    will be incorporated in the listing. The default is 'ignore'.

    Only the first parameter is mandatory.

     # list context
     foreach my $file (parse_dir($listing)) {
         my($name, $type, $size, $mtime, $mode) = @$file;
     }
 
     # scalar context
     my $dir = parse_dir($listing);
     foreach my $file (@$dir) {
         my($name, $type, $size, $mtime, $mode) = @$file;
     }

    The return value from parse_dir() is a list of directory entries. In a
    scalar context the return value is a reference to the list. The
    directory entries are represented by an array consisting of:

    name
        The name of the file.

    type
        One of: "f" file, "d" directory, "l" symlink, "?" unknown.

    size
        The size of the file.

    time
        The number of seconds since January 1, 1970.

    mode
        Bitmask a la the mode returned by "stat".

SEE ALSO
    File::Listing::Ftpcopy
        Provides the same interface but uses XS and the parser
        implementation from "ftpcopy".

COPYRIGHT
    Copyright 1996-2010, Gisle Aas Copyright 2020 Graham Ollis

    Based on lsparse.pl (from Lee McLoughlin's ftp mirror package) and
    Net::FTP's parse_dir (Graham Barr).

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

6.16      2023-07-12 15:22:25 -0600
  - Support dosftp listings with four-digit years (gh#3, gh#26)

6.15      2022-04-17 04:09:58 -0600
  - Main git repository has now been detached from the original
    repository (originally it was a fork).  The old repository
    can be found at https://github.com/gisle/file-listing
  - Adjust test suite to handle systems that cannot handle
    year 2038+ dates (gh#24)

6.14      2020-11-30 05:48:07 -0700
  - Production version identical to 6.12_01

6.12_01   2020-11-27 16:13:15 -0700
  - Handle default apache indexes (gh#20)
  - Ignore certain known navigation links in apache indexes (gh#20)
  - Support Win32 OpenSSH listing (gh#21 simon04++)

6.11      2020-10-06 03:00:45 -0600
  - Remove accidental recursive req on LWP::Simple

6.10      2020-10-05 15:18:30 -0600
  - Production version identical to 6.09_01

6.09_01   2020-10-05 08:17:32 -0600
  - Remove TABs from source files and tests (gh#17)
  - Restores (theoretical, untested) support for 5.6, which was removed when
    the tests added modern file tests (gh#17)
  - Fixes or silences several minor nits reported by PerlCritic (gh#17)

6.08_01 2020-10-01 17:28:35 -0600
  - We now only call Time::Local::timelocal with four digit years (gh#5, gh#14)

6.07 2020-09-30 08:47:18 -0600
  - Production version identical to 6.06_01

6.06_01 2020-09-30 06:34:08 -0600
  - Add examples to documentation.

6.05_01 2020-09-29 21:25:35 -0600
  - New maintainer (Graham Ollis / plicease)
  - New GitHub repository https://github.com/PerlAlien/File-Listing
  - Include Perl core modules in prereqs (rt#85100 / gh#2)
  - Converted Changes file to standard format (the original
    Changes file is available as Changes.original)
  - Migrate Test.pm tests to use Test::More, all tests
    use strict and warnings now.

6.04 2012-02-15
  - Minor mistake in module description (Erik Esterer)
  - Test parsing of dosftp listing (Gisle Aas)
  - This module works with 5.6.2 (Peter Rabbitson)

6.03 2011-09-20
  - Minor mistake in module description (Erik Esterer)
  - Fixes to the parsing of 'ls -l' file modes (rt#71054, Peter John Acklam)

6.02 2011-03-20
  - Avoid running live test.  Embed the expected test data in  the test file.
    Update the test data inline by invoking 'perl t/apache.t --update'.

6.01 2011-03-13
  - Added legal notice and updated the meta repository link

6.00 2011-02-25
  - Initial release of File-Listing as a separate  distribution.  The File::Listing
    module used to be bundled with the libwww-perl distribution.

  - Since libwww-perl-5.837 File::Listing has been made able to deal with listings
    full of table markup (like what we find at http://www.cpan.org/modules/by-module/).


Makefile.PL
#!perl -w

use 5.006002;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME => 'File::Listing',
    VERSION_FROM => 'lib/File/Listing.pm',
    ABSTRACT_FROM => 'lib/File/Listing.pm',
    AUTHOR => 'Gisle Aas <gisle@activestate.com>',
    LICENSE => "perl",
    MIN_PERL_VERSION => 5.006002,
    PREREQ_PM => {
	'HTTP::Date'  => 6,
	'Exporter'    => 0,
	'Time::Local' => 0,
	'base'        => 0,
    },
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    TEST_REQUIRES => {
        'Test::More' => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            bugtracker => {
                web  => 'https://github.com/PerlAlien/File-Listing/issues',
            },
            homepage => 'https://metacpan.org/pod/File::Listing',
            repository => {
                type => 'git',
                url  => 'https://github.com/PerlAlien/File-Listing.git',
                web  => 'https://github.com/PerlAlien/File-Listing',
            },
        },
    },
);


BEGIN {
    # compatibility with older versions of MakeMaker
    my $developer = -f ".gitignore";
    my %mm_req = (
        LICENCE => 6.31,
        META_MERGE => 6.45,
        META_ADD => 6.45,
        MIN_PERL_VERSION => 6.48,
        CONFIGURE_REQUIRES => 6.52,
        TEST_REQUIRES => 6.64,
    );
    undef(*WriteMakefile);
    *WriteMakefile = sub {
        my %arg = @_;
        for (keys %mm_req) {
            unless (eval { ExtUtils::MakeMaker->VERSION($mm_req{$_}) }) {
                warn "$_ $@" if $developer;
                delete $arg{$_};
            }
        }
        ExtUtils::MakeMaker::WriteMakefile(%arg);
    };
}


lib/File/Listing.pm
package File::Listing;

use strict;
use warnings;
use Carp ();
use HTTP::Date qw(str2time);
use Exporter 5.57 qw( import );

# ABSTRACT: Parse directory listing
# VERSION

sub Version { $File::Listing::VERSION; }

our @EXPORT = qw(parse_dir);

sub parse_dir ($;$$$)
{
   my($dir, $tz, $fstype, $error) = @_;

   $fstype ||= 'unix';
   $fstype = "File::Listing::" . lc $fstype;

   my @args = $_[0];
   push(@args, $tz) if(@_ >= 2);
   push(@args, $error) if(@_ >= 4);

   $fstype->parse(@args);
}


sub line { Carp::croak("Not implemented yet"); }
sub init { } # Dummy sub


sub file_mode ($)
{
    Carp::croak("Input to file_mode() must be a 10 character string.")
        unless length($_[0]) == 10;

    # This routine was originally borrowed from Graham Barr's
    # Net::FTP package.

    local $_ = shift;
    my $mode = 0;
    my($type);

    s/^(.)// and $type = $1;

    # When the set-group-ID bit (file mode bit 02000) is set, and the group
    # execution bit (file mode bit 00020) is unset, and it is a regular file,
    # some implementations of `ls' use the letter `S', others use `l' or `L'.
    # Convert this `S'.

    s/[Ll](...)$/S$1/;

    while (/(.)/g) {
        $mode <<= 1;
        $mode |= 1 if $1 ne "-" &&
                      $1 ne "*" &&
                      $1 ne 'S' &&
                      $1 ne 'T';
    }

    $mode |= 0004000 if /^..s....../i;
    $mode |= 0002000 if /^.....s.../i;
    $mode |= 0001000 if /^........t/i;

    # De facto standard definitions. From 'stat.h' on Solaris 9.

    $type eq "p" and $mode |= 0010000 or        # fifo
    $type eq "c" and $mode |= 0020000 or        # character special
    $type eq "d" and $mode |= 0040000 or        # directory
    $type eq "b" and $mode |= 0060000 or        # block special
    $type eq "-" and $mode |= 0100000 or        # regular
    $type eq "l" and $mode |= 0120000 or        # symbolic link
    $type eq "s" and $mode |= 0140000 or        # socket
    $type eq "D" and $mode |= 0150000 or        # door
      Carp::croak("Unknown file type: $type");

    $mode;
}


sub parse
{
   my($pkg, $dir, $tz, $error) = @_;

   # First let's try to determine what kind of dir parameter we have
   # received.  We allow both listings, reference to arrays and
   # file handles to read from.

   if (ref($dir) eq 'ARRAY') {
       # Already split up
   }
   elsif (ref($dir) eq 'GLOB') {
       # A file handle
   }
   elsif (ref($dir)) {
      Carp::croak("Illegal argument to parse_dir()");
   }
   elsif ($dir =~ /^\*\w+(::\w+)+$/) {
      # This scalar looks like a file handle, so we assume it is
   }
   else {
      # A normal scalar listing
      $dir = [ split(/\n/, $dir) ];
   }

   $pkg->init();

   my @files = ();
   if (ref($dir) eq 'ARRAY') {
       for (@$dir) {
           push(@files, $pkg->line($_, $tz, $error));
       }
   }
   else {
       local($_);
       while (my $line = <$dir>) {
           chomp $line;
           push(@files, $pkg->line($line, $tz, $error));
       }
   }
   wantarray ? @files : \@files;  ## no critic (Community::Wantarray)
}



package File::Listing::unix;

use HTTP::Date qw(str2time);

our @ISA = qw(File::Listing);

# A place to remember current directory from last line parsed.
our $curdir;

sub init
{
    $curdir = '';
}


sub line
{
    shift; # package name
    local($_) = shift;
    my($tz, $error) = @_;

    s/\015//g;
    #study;

    my ($kind, $size, $date, $name);
    if (($kind, $size, $date, $name) =
        /^([\-\*FlrwxsStTdD]{10})                 # Type and permission bits
         .*                                       # Graps
         \D(\d+)                                  # File size
         \s+                                      # Some space
         (\w{3}\s+\d+\s+(?:\d{1,2}:\d{2}|\d{4})|\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2})  # Date
         \s+                                      # Some more space
         (.*)$                                    # File name
        /x )

    {
        return if $name eq '.' || $name eq '..';
        $name = "$curdir/$name" if length $curdir;
        my $type = '?';
        if ($kind =~ /^l/ && $name =~ /(.*) -> (.*)/ ) {
            $name = $1;
            $type = "l $2";
        }
        elsif ($kind =~ /^[\-F]/) { # (hopefully) a regular file
            $type = 'f';
        }
        elsif ($kind =~ /^[dD]/) {
            $type = 'd';
            $size = undef;  # Don't believe the reported size
        }
        return [$name, $type, $size, str2time($date, $tz),
              File::Listing::file_mode($kind)];

    }
    elsif (/^(.+):$/ && !/^[dcbsp].*\s.*\s.*:$/ ) {
        my $dir = $1;
        return () if $dir eq '.';
        $curdir = $dir;
        return ();
    }
    elsif (/^[Tt]otal\s+(\d+)$/ || /^\s*$/) {
        return ();
    }
    elsif (/not found/    || # OSF1, HPUX, and SunOS return
             # "$file not found"
             /No such file/ || # IRIX returns
             # "UX:ls: ERROR: Cannot access $file: No such file or directory"
                               # Solaris returns
             # "$file: No such file or directory"
             /cannot find/     # Windows NT returns
             # "The system cannot find the path specified."
             ) {
        return () unless defined $error;
        &$error($_) if ref($error) eq 'CODE';
        warn "Error: $_\n" if $error eq 'warn';
        return ();
    }
    elsif ($_ eq '') {       # AIX, and Linux return nothing
        return () unless defined $error;
        &$error("No such file or directory") if ref($error) eq 'CODE';
        warn "Warning: No such file or directory\n" if $error eq 'warn';
        return ();
    }
    else {
        # parse failed, check if the dosftp parse understands it
        File::Listing::dosftp->init();
        return(File::Listing::dosftp->line($_,$tz,$error));
    }

}



package File::Listing::dosftp;

use HTTP::Date qw(str2time);

our @ISA = qw(File::Listing);

# A place to remember current directory from last line parsed.
our $curdir;



sub init
{
    $curdir = '';
}


sub line
{
    shift; # package name
    local($_) = shift;
    my($tz, $error) = @_;

    s/\015//g;

    my ($date, $size_or_dir, $name, $size);

    # usual format:
    # 02-05-96  10:48AM                 1415 src.slf
    # 09-10-96  09:18AM       <DIR>          sl_util
    # alternative dos format with four-digit year:
    # 02-05-2022  10:48AM                 1415 src.slf
    # 09-10-2022  09:18AM       <DIR>          sl_util
    if (($date, $size_or_dir, $name) =
        /^(\d\d-\d\d-\d{2,4}\s+\d\d:\d\d\wM)      # Date and time info
         \s+                                      # Some space
         (<\w{3}>|\d+)                            # Dir or Size
         \s+                                      # Some more space
         (.+)$                                    # File name
        /x )
    {
        return if $name eq '.' || $name eq '..';
        $name = "$curdir/$name" if length $curdir;
        my $type = '?';
        if ($size_or_dir eq '<DIR>') {
            $type = "d";
            $size = ""; # directories have no size in the pc listing
        }
        else {
            $type = 'f';
            $size = $size_or_dir;
        }
        return [$name, $type, $size, str2time($date, $tz), undef];
    }
    else {
        return () unless defined $error;
        &$error($_) if ref($error) eq 'CODE';
        warn "Can't parse: $_\n" if $error eq 'warn';
        return ();
    }

}



package File::Listing::vms;
our @ISA = qw(File::Listing);

package File::Listing::netware;
our @ISA = qw(File::Listing);



package File::Listing::apache;

our @ISA = qw(File::Listing);


sub init { }


sub line {
    shift; # package name
    local($_) = shift;
    my($tz, $error) = @_; # ignored for now...

    s!</?t[rd][^>]*>! !g;  # clean away various table stuff
    if (m!<A\s+HREF=\"([^?\"]+)\">.*</A>.*?(\d+)-([a-zA-Z]+|\d+)-(\d+)\s+(\d+):(\d+)\s+(?:([\d\.]+[kMG]?|-))!i) {
        my($filename, $filesize) = ($1, $7);
        my($d,$m,$y, $H,$M) = ($2,$3,$4,$5,$6);
        if ($m =~ /^\d+$/) {
            ($d,$y) = ($y,$d) # iso date
        }
        else {
            $m = _monthabbrev_number($m);
        }

        $filesize = 0 if $filesize eq '-';
        if ($filesize =~ s/k$//i) {
            $filesize *= 1024;
        }
        elsif ($filesize =~ s/M$//) {
            $filesize *= 1024*1024;
        }
        elsif ($filesize =~ s/G$//) {
            $filesize *= 1024*1024*1024;
        }
        $filesize = int $filesize;

        require Time::Local;
        my $filetime = Time::Local::timelocal(0,$M,$H,$d,$m-1,_guess_year($y));
        my $filetype = ($filename =~ s|/$|| ? "d" : "f");
        return [$filename, $filetype, $filesize, $filetime, undef];
    }

    # the default listing doesn't include timestamps or file sizes
    # but we don't want to grab navigation links, so we ignore links
    # that have a non-trailing slash / character or ?
    elsif(m!<A\s+HREF=\"([^?/\"]+/?)\">.*</A>!i) {
        my $filename = $1;
        my $filetype = ($filename =~ s|/$|| ? "d" : "f");
        return [$filename, $filetype, undef, undef, undef];
    }

    return ();
}


sub _guess_year {
    my $y = shift;

    # if the year is already four digit then we shouldn't do
    # anything to modify it.
    if ($y >= 1900) {
        # do nothing

    # TODO: for hysterical er historical reasons we assume 9x is in the
    # 1990s we should probably not do that, but I don't have any examples
    # where apache provides two digit dates so I am leaving this as-is
    # for now.  Possibly the right thing is to not handle two digit years.
    } elsif ($y >= 90) {
        $y = 1900+$y;
    }

    # TODO: likewise assuming 00-89 are 20xx is long term probably wrong.
    elsif ($y < 100) {
        $y = 2000+$y;
    }
    $y;
}


sub _monthabbrev_number {
    my $mon = shift;
    +{'Jan' => 1,
      'Feb' => 2,
      'Mar' => 3,
      'Apr' => 4,
      'May' => 5,
      'Jun' => 6,
      'Jul' => 7,
      'Aug' => 8,
      'Sep' => 9,
      'Oct' => 10,
      'Nov' => 11,
      'Dec' => 12,
     }->{$mon};
}


1;

__END__

=head1 SYNOPSIS

 use File::Listing qw(parse_dir);
 $ENV{LANG} = "C";  # dates in non-English locales not supported
 foreach my $file (parse_dir(`ls -l`)) {
     my ($name, $type, $size, $mtime, $mode) = @$file;
     next if $type ne 'f'; # plain file
     #...
 }
 
 # directory listing can also be read from a file
 open my $listing, "zcat ls-lR.gz|";
 $dir = parse_dir($listing, '+0000');

=head1 DESCRIPTION

This module exports a single function called C<parse_dir>, which can be
used to parse directory listings.

=head1 FUNCTIONS

=head2 parse_dir

 my $dir = parse_dir( $listing );
 my $dir = parse_dir( $listing, $time_zone );
 my $dir = parse_dir( $listing, $time_zone, $type );
 my $dir = parse_dir( $listing, $time_zone, $type, $error );
 my @files = parse_dir( $listing );
 my @files = parse_dir( $listing, $time_zone );
 my @files = parse_dir( $listing, $time_zone, $type );
 my @files = parse_dir( $listing, $time_zone, $type, $error );

The first parameter (C<$listing>) is the directory listing to parse.
It can be a scalar, a reference to an array of directory lines or a
glob representing a filehandle to read the directory listing from.

The second parameter (C<$time_zone>) is the time zone to use when
parsing time stamps in the listing. If this value is undefined,
then the local time zone is assumed.

The third parameter (C<$type>) is the type of listing to assume.
Currently supported formats are C<'unix'>, C<'apache'> and
C<'dosftp'>. The default value is C<'unix'>. Ideally, the listing
type should be determined automatically.

The fourth parameter (C<$error>) specifies how unparseable lines
should be treated. Values can be C<'ignore'>, C<'warn'> or a code reference.
Warn means that the perl warn() function will be called.  If a code
reference is passed, then this routine will be called and the return
value from it will be incorporated in the listing.  The default is
C<'ignore'>.

Only the first parameter is mandatory.

 # list context
 foreach my $file (parse_dir($listing)) {
     my($name, $type, $size, $mtime, $mode) = @$file;
 }
 
 # scalar context
 my $dir = parse_dir($listing);
 foreach my $file (@$dir) {
     my($name, $type, $size, $mtime, $mode) = @$file;
 }

The return value from parse_dir() is a list of directory entries.
In a scalar context the return value is a reference to the list.
The directory entries are represented by an array consisting of:

=over 4

=item name

The name of the file.

=item type

One of: C<f> file, C<d> directory, C<l> symlink, C<?> unknown.

=item size

The size of the file.

=item time

The number of seconds since January 1, 1970.

=item mode

Bitmask a la the mode returned by C<stat>.

=back

=head1 SEE ALSO

=over 4

=item L<File::Listing::Ftpcopy>

Provides the same interface but uses XS and the parser implementation from C<ftpcopy>.

=back

=cut
